### DISCUSSION TITLE
great tables with quarto and typst

### DISCUSSION THREAD

**Discussion by boshek** [quarto-dev/quarto-cli#10516] [Q&A] [answered]
### Description

I have noticed some differences between great tables output versus that which I get when I use great tables with typst and quarto. I have just started using typst so these could very well be skill issues. Additionally I am opening this first here but feel free to suggest if I should open these in the great tables repo. 

I've written a small qmd that replicates what I am noticing. My approach is to create a table but also write on an artifact. When replicated with `format:html` the table renders as expected. Here is what is generated from the great tables code when I save them as pngs. This is the expected behaviour:

## cell_borders

<img width="259" alt="image" src="https://github.com/user-attachments/assets/be455b58-4e65-4120-a1a6-137e6c9ea46c">

## italics

<img width="266" alt="image" src="https://github.com/user-attachments/assets/23e75a44-fbb1-4671-9739-379167d10c16">

## bold

<img width="260" alt="image" src="https://github.com/user-attachments/assets/81ad1519-b373-488e-98fb-53262da17377">

And then here is what the qmd looks like as a pdf when rendered with typst:

<img width="363" alt="image" src="https://github.com/user-attachments/assets/adaf8a91-6d57-4926-a539-2fec988b616c">

I was not expecting 100% replication of a PDF table to an HTML table but my assumption was that styles would be implemented. Sorry I'm not familiar enough with typst to resolve the HTML to typst process (though I'm happy to receive any tips!)

````qmd
---
format: typst
execute: 
  echo: false
---

```{python}
from great_tables import GT, style, loc, exibble
exibble_sm = exibble[["num", "fctr", "row", "group"]].head(5)
gt_tbl = GT(exibble_sm)
gt_tbl
```


## cell borders

```{python}
cell_borders = (
    GT(exibble_sm)   
    .tab_style(
        style=style.borders(sides=["bottom"], weight='2px', color="red"),
        locations=loc.body(rows=[3])
    ) 
)
cell_borders
```

```{python}
cell_borders.save("cell_borders.png")
```

## italics

```{python}
italics = (
    GT(exibble_sm)   
    .tab_style(
        style=style.text(style="italic"),
        locations=loc.body(rows=[3])
    ) 
)
italics
```

```{python}
italics.save("italics.png")
```

## bold

```{python}
bold = (
    GT(exibble_sm)   
    .tab_style(
        style=style.text(weight="bold"),
        locations=loc.body(rows=[3])
    ) 
)
bold
```

```{python}
bold.save("bold.png")
```
````

**Comment by cscheid** [comment]
Those kinds of differences are unavoidable, because the rendering engines are not precisely the same. The idea of this feature is not for tables to come out looking identical: it exists to empower you to change the styles in a similar way as you would in HTML.

**Reply by boshek** [reply]
Gotcha. Thanks for this response.

**Reply by cscheid** [reply]
I should have read a bit further, though, and I think we might be able to improve the bold/italic/etc rendering in the future - but padding, outlines, and precise placement will never quite perfectly match.

And, I'm not sure how to explain the differences, but this is what I see:

<img width="325" alt="image" src="https://github.com/user-attachments/assets/50569cf6-fb94-4df4-b986-aae06feb46e7">

**Reply by aito123** [reply]
Good to hear that this issue is solvable. Is this bug-fix going to be present in the comming updates of quarto? That would be really great. Thanks.

**Comment by aito123** [comment]
Hello,
I can confirm that this behaviour is also present in R `{gt}` package. I'm using Quarto 1.7 and Typst 0.13

This is the qmd reprex:

````{md}
---
title: "test1"
format: typst
---

## gt tables

```{r}
pacman::p_load(gt, tidyverse)
```

### border

```{r}
exibble %>% 
  select(1:3) %>% 
  gt() %>% 
  tab_style(style = list(cell_borders(color = "red")),
              locations = cells_body(rows = char %in% "apricot"))

```


### italics

```{r}
exibble %>% 
  select(1:3) %>% 
  gt() %>% 
  tab_style(style = list(cell_text(style = "italic")),
              locations = cells_body(rows = char %in% "apricot"))
```


### bold

```{r}
exibble %>% 
  select(1:3) %>% 
  gt() %>% 
  tab_style(style = list(cell_text(weight = "bold")),
              locations = cells_body(rows = char %in% "apricot"))
```

````

Tables should look like this:
![image](https://github.com/user-attachments/assets/8bcd2317-0ad2-4baa-9ded-2d64df41dac8)
![image](https://github.com/user-attachments/assets/7b74f5eb-21f8-410c-a407-3a8973ca6480)
![image](https://github.com/user-attachments/assets/38828746-fd16-4131-9b57-105dd1a1a02b)

But look like this:
![image](https://github.com/user-attachments/assets/0b6bfd72-4604-4f58-bd7c-24bf4c7d61ee)
![image](https://github.com/user-attachments/assets/90c96426-749d-40e4-81c0-df2869788754)

<details>
<summary> Quarto info </summary>

Quarto 1.7.21
[>] Checking environment information...
      Quarto cache location: D:\Users\santiago.sotelo\AppData\Local\quarto
[>] Checking versions of quarto binary dependencies...
      Pandoc version 3.6.3: OK
      Dart Sass version 1.85.1: OK
      Deno version 1.46.3: OK
      Typst version 0.13.0: OK
[>] Checking versions of quarto dependencies......OK
[>] Checking Quarto installation......OK
      Version: 1.7.21
      Path: D:\Users\santiago.sotelo\AppData\Local\Programs\Quarto\bin
      CodePage: 1252

[>] Checking tools....................OK
      TinyTeX: v2024.11
      Chromium: (not installed)

[>] Checking LaTeX....................OK
      Using: TinyTex
      Path: D:\Users\santiago.sotelo\AppData\Roaming\TinyTeX\bin\windows\
      Version: 2024

[>] Checking Chrome Headless....................OK
      Using: Chrome found on system
      Path: C:\Program Files\Google\Chrome\Application\chrome.exe
      Source: Windows Registry

[>] Checking basic markdown render....OK

[>] Checking Python 3 installation....OK
      Version: 3.11.12
      Path: D:/Users/santiago.sotelo/AppData/Local/R/cache/R/reticulate/uv/cache/archive-v0/vHld6hSKmvUV1AGM3oRJl/Scripts/python.exe
      Jupyter: (None)

      Jupyter is not available in this Python installation.
      Install with py -m pip install jupyter

[>] Checking R installation...........OK
      Version: 4.4.2
      Path: D:/Users/santiago.sotelo/AppData/Local/Programs/R/R-4.4.2
      LibPaths:
        - D:/Users/santiago.sotelo/AppData/Local/Programs/R/R-4.4.2/library
      knitr: 1.50
      rmarkdown: 2.29

[>] Checking Knitr engine render......OK

</details>

**Reply by cderv** [reply]
@gordonwoodhull Are these known limitations for CSS translation support in Typst? Do we need to log an issue about them?

**Reply by gordonwoodhull** [reply, marked as answer]
Thanks for the heads-up! Not sure how I missed this discussion the first time around.

`font-weight` and `font-style` are not implemented, but I did a quick experiment and they easily could be. These are pretty important!

In general, [this table in the advanced docs](https://quarto.org/docs/advanced/typst/typst-css.html#supported-elements-and-properties) summarizes what element/property combinations are supported.

**Reply by cderv** [reply]
Awesome! I let you open an issue with your experiment findings so that we don't forget that for 1.8

**Reply by cderv** [reply]
In fact this is here: 
- quarto-dev/quarto-cli/pull/12554

**Reply by gordonwoodhull** [reply]
Thanks! It’s annoying that GH doesn’t create backlinks for discussions when you refer to them, the same way it does for issues.

**Reply by cderv** [reply]
Yes definitely. It creates some links if you create an issue from a discussion, but its rarely done this way; Hopefully they'll do an update
